components/script/animations.rs

```diff
@@ -6,141 +6,175 @@
 }
 
 impl Animations {
     pub(crate) fn new() -> Self {
         Animations {
             sets: Default::default(),
-            have_running_animations: false,
+            have_running_animations: DomRefCell::new(false),
+            rooted_nodes: Default::default(),
+            pending_events: Default::default(),
         }
     }
 
-    pub(crate) fn update_for_new_timeline_value(
-        &mut self,
-        window: &Window,
-        now: f64,
-    ) -> AnimationsUpdate {
-        let mut update = AnimationsUpdate::new(window.pipeline_id());
+    pub(crate) fn clear(&self) {
+        self.sets.write().clear();
+        self.rooted_nodes.borrow_mut().clear();
+        self.pending_events.borrow_mut().clear();
+    }
+
+    pub(crate) fn mark_animating_nodes_as_dirty(&self) {
+        let sets = self.sets.read();
+        let rooted_nodes = self.rooted_nodes.borrow();
+        for node in sets.keys().filter_map(|node| rooted_nodes.get(&node)) {
+            node.dirty(NodeDamage::NodeStyleDamaged);
+        }
+    }
+
+    pub(crate) fn update_for_new_timeline_value(&self, window: &Window, now: f64) {
+        let pipeline_id = window.pipeline_id();
         let mut sets = self.sets.write();
 
         for set in sets.values_mut() {
             // When necessary, iterate our running animations to the next iteration.
             for animation in set.animations.iter_mut() {
                 if animation.iterate_if_necessary(now) {
-                    update.add_event(
-                        animation.node,
-                        animation.name.to_string(),
+                    self.add_animation_event(
+                        animation,
                         TransitionOrAnimationEventType::AnimationIteration,
-                        animation.active_duration(),
+                        pipeline_id,
                     );
                 }
             }
 
-            Self::finish_running_animations(set, now, &mut update);
+            self.finish_running_animations(set, now, pipeline_id);
         }
-        update
+
+        self.unroot_unused_nodes(&sets);
     }
 
     /// Processes any new animations that were discovered after reflow. Collect messages
     /// that trigger events for any animations that changed state.
     /// TODO(mrobinson): The specification dictates that this should happen before reflow.
-    pub(crate) fn do_post_reflow_update(&mut self, window: &Window, now: f64) -> AnimationsUpdate {
-        let mut update = AnimationsUpdate::new(window.pipeline_id());
-
-        {
-            let mut sets = self.sets.write();
-            update.collect_newly_animating_nodes(&sets);
-
-            for set in sets.values_mut() {
-                Self::handle_canceled_animations(set, now, &mut update);
-                Self::handle_new_animations(set, &mut update);
-            }
+    pub(crate) fn do_post_reflow_update(
+        &self,
+        window: &Window,
+        js_runtime: *mut JSRuntime,
```

Comment:
Let's take a &Runtime argument here, rather than a raw pointer. We could even add a getter to Window and remove the argument.

Reply:
Okay. This makes things quite a bit simpler, so I'll add a getter and pass around `&Window` everywhere.